src/util.cpp

```diff
@@ -394,6 +394,44 @@ void AdaptiveScheduler::reset() {
   start_time = std::chrono::steady_clock::now();
 }
 
+ProgressMonitor::ProgressMonitor(int64_t target_seconds,
+                                 const std::string &progress_file)
+    : start_time(std::chrono::steady_clock::now()),
+      target_seconds(target_seconds),
+      progress_file(progress_file) {
+  if (target_seconds <= 0) {
+    throw std::runtime_error("internal error: invalid target duration: " +
+                             std::to_string(target_seconds));
+  }
+}
+
+int64_t ProgressMonitor::getElapsedSeconds() {
+  const auto now = std::chrono::steady_clock::now();
+  return std::chrono::duration_cast<std::chrono::seconds>(now - start_time)
+      .count();
+}
+
+bool ProgressMonitor::isTargetReached() {
+  return getElapsedSeconds() >= target_seconds;
+}
+
+double ProgressMonitor::getProgress() {
+  double progress = static_cast<double>(getElapsedSeconds()) /
+                    static_cast<double>(target_seconds);
+  if (progress > 1.0) {
+    progress = 1.0;
+  }
```

Comment:
If the elapsed seconds return with a negative value, then perhaps `if (progress < 0.0) { progress = 0.0; }`